File: notifier/src/main/java/com/event/processing/notifier/service/WebhookService.java
package com.event.processing.notifier.service;

import com.event.processing.notifier.domain.dto.BaseEventDTO;
import com.event.processing.notifier.domain.dto.WebhookEventDTO;

/**
 * Interface defining the contract for webhook notification processing.
 * This interface provides a standardized way to process webhook notifications
 * with retry capabilities for improved reliability.
 * <p>
 * Key features:
 * - Retry mechanism for failed webhook deliveries
 * - Event payload processing
 * - Webhook URL handling
 * - Event identification
 *
 * @author LongLe
 * @version 1.0
 */
public interface WebhookService {
  /**
   * Processes a webhook notification with retry capabilities.
   * This method handles the delivery of webhook notifications to the specified
   * URL,
   * implementing retry logic for failed attempts.
   *
   * @param eventId      The unique identifier of the event being processed
   * @param eventPayload The webhook event payload containing event details
   * @param url          The destination URL for the webhook notification
   * @param payload      The base event payload to be sent in the webhook
   */
  void processWithRetry(String eventId, WebhookEventDTO eventPayload, String url, BaseEventDTO payload);
}

File: notifier-go/service/webhook_service.go
package service

import (
	"context"

	"github.com/event-processing/notifier-go/domain/dto"
)

// WebhookService defines the interface for webhook notification processing
type WebhookService interface {
	// ProcessWithRetry processes a webhook notification with retry capabilities
	ProcessWithRetry(ctx context.Context, eventID string, eventPayload *dto.WebhookEventDTO, url string, payload dto.BaseEventDTO) error
}

// WebhookClient defines the interface for webhook delivery functionality
type WebhookClient interface {
	// SendWebhook sends a webhook notification to the specified URL
	SendWebhook(ctx context.Context, url string, payload dto.BaseEventDTO) (bool, error)
}

// DeadLetterQueueProducer defines the interface for publishing failed events to a dead letter queue
type DeadLetterQueueProducer interface {
	// Publish publishes a failed event to the dead letter queue
	Publish(ctx context.Context, topic, key string, payload *dto.DeadLetterQueueEventDTO) error
}

// MetricsService defines the interface for metrics collection
type MetricsService interface {
	// IncrementWebhookSuccess increments the webhook success counter
	IncrementWebhookSuccess()

	// IncrementWebhookFailure increments the webhook failure counter
	IncrementWebhookFailure()

	// IncrementCircuitBreakerOpenCount increments the circuit breaker open counter
	IncrementCircuitBreakerOpenCount()
}

